cbmc/proofs/mqtt_state.c

```diff
@@ -138,38 +138,43 @@ IotListDouble_t *allocate_IotMqttOperationList( IotListDouble_t *pOp,
 						size_t length,
 						IotMqttConnection_t pConn )
 {
-  if ( pOp == NULL ) pOp = malloc_can_fail( sizeof( *pOp ) );
-  if ( pOp == NULL ) return NULL;
-
   // Allocate lists of length L <= 3 (MAX = L+1)
-  __CPROVER_assert(OPERATION_COUNT_MAX <= 3+1,
-		   "Operation list bound is too big");
   __CPROVER_assert(length < OPERATION_COUNT_MAX,
-		   "Operation list requested is too long");
+		   "Operation list length max is greater than MAX");
+  __CPROVER_assert(length <= 3,
+		   "Operation list length max is greater than 3");
+
+  if ( pOp == NULL ) pOp = malloc_can_fail( sizeof( *pOp ) );
+  if ( pOp == NULL ) return NULL;
 
   IotListDouble_Create( pOp );
 
   size_t num_elts;
   __CPROVER_assume(num_elts <= length);
 
-  if (1 <= num_elts)
-    {
-      IotMqttOperation_t pElt = allocate_IotMqttOperation( NULL, pConn );
-      __CPROVER_assume( pElt );
-      IotListDouble_InsertHead( pOp, &( pElt->link ) );
-    }
-  if (2 <= num_elts)
-    {
-      IotMqttOperation_t pElt = allocate_IotMqttOperation( NULL, pConn );
-      __CPROVER_assume( pElt );
-      IotListDouble_InsertHead( pOp, &( pElt->link ) );
-    }
-  if (3 <= num_elts)
-    {
-      IotMqttOperation_t pElt = allocate_IotMqttOperation( NULL, pConn );
-      __CPROVER_assume( pElt );
-      IotListDouble_InsertHead( pOp, &( pElt->link ) );
-    }
+  IotMqttOperation_t pElt;
+  switch (num_elts) {
```

Comment:
@aggarg This code is not designed to be executed or compiled, but to be verified with CBMC. Therefore, we must avoid loops as much as possible, since they dramatically decrease CBMC performance. In this  case, we are assuming lists are of length at most `OPERATION_COUNT_MAX`. This code is nothing but the straightline code obtained by unrolling the loop OPERATION_COUNT_MAX iterations.